test/integration/replicaschedulingpreference_test.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2018 The Federation v2 Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"testing"
+
+	fedschedulingv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/federatedscheduling/v1alpha1"
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/federation/v1alpha1"
+	"github.com/kubernetes-sigs/federation-v2/pkg/client/clientset_generated/clientset"
+	"github.com/kubernetes-sigs/federation-v2/test/common"
+	"github.com/kubernetes-sigs/federation-v2/test/integration/framework"
+	appsv1 "k8s.io/api/apps/v1"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// TestReplicaSchedulingPreference validates basic replica scheduling preference calculations.
+var TestReplicaSchedulingPreference = func(t *testing.T) {
+	t.Parallel()
+	tl := framework.NewIntegrationLogger(t)
+
+	controllerFixture, fedClient := initRSPTest(tl, FedFixture)
+	defer controllerFixture.TearDown(tl)
+
+	clusters := getClusterNames(fedClient)
+	if len(clusters) != 2 {
```

Comment:
'< 2' ?

Reply:
Some of the tabular tests here expect exactly 2 clusters. This could be a difference number also, if the tests cases need that (may be in future).

Reply:
(No action required) Requiring such specific setup is a smell to me.  I think integration testing would ideally be limited to retrieving details from an arbitrary number of member clusters and the federation api, and ensuring the resulting computation is applied correctly to the correct resources.  The specifics tested here that depend on an exact number of clusters would seem to be easier to evaluate in a unit test that provided specific inputs without the overhead of fixture.  I'm not suggesting this gets fixed here, but I anticipate this guideline being usefully applied in the future.